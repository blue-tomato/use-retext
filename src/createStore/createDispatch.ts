import get from 'just-safe-get';
import set from 'just-safe-set';
import mitt from 'mitt';
import { mapValuesDeep } from '../helpers';
import { Reducer } from '../Store';

interface ReducerInfo {
  scope: string;
  reducer: Reducer;
}

export default (initialState: {}, actions: {}): [any, mitt.Emitter] => {
  const groupedReducer: Record<string, Array<ReducerInfo>> = {};
  const emitter = mitt();
  let currentState = initialState;

  return [
    mapValuesDeep(actions, ({ key, value, scope: actionScope }) => {
      if (!groupedReducer[key]) {
        groupedReducer[key] = [];
      }

      groupedReducer[key].push({
        scope: actionScope,
        reducer: (value as Function)(),
      });

      return (payload?: unknown) => {
        groupedReducer[key].forEach(({ scope, reducer }: ReducerInfo) => {
          // Ensure that the scope starts at least with the action-scope
          if (!scope.startsWith(actionScope)) return;

          const state = scope ? get(currentState, scope) : currentState;
          const newState = reducer(state, payload);

          if (scope) {
            set(currentState, scope, {
              ...get(currentState, scope),
              ...newState,
            });
          } else {
            currentState = {
              ...currentState,
              ...newState,
            };
          }
        });

        emitter.emit('update', {
          newState: currentState,
          scope: actionScope,
        });

        // return currentState;
      };
    }),
    emitter,
  ];
};
