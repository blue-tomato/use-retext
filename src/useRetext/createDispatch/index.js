import mitt from 'mitt';
import { mapValuesDeep } from '../../helpers';

export default action => {
  const emitter = mitt();
  const dispatch = mapValuesDeep(action, ({ key, scope }) => payload => {
    emitter.emit(key, { scope, payload });
  });

  return [dispatch, emitter];
};
