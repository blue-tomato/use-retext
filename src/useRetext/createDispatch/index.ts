import mitt from 'mitt';
import { mapValuesDeep } from '../../helpers';
import { Action } from '../../Store';

export default <S>(action: Action<S>) => {
  const emitter = mitt();
  const dispatch = mapValuesDeep(action, ({ key, scope }) => (payload: any) => {
    emitter.emit(key, { scope, payload });
  });

  return [dispatch, emitter];
};
