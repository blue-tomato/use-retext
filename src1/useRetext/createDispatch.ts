import mitt from 'mitt';
import { mapValuesDeep } from '../helpers';

export default (action: object): [unknown, mitt.Emitter] => {
  const emitter = mitt();
  const dispatch = mapValuesDeep(action, ({ key, scope }) => (payload?: unknown) =>
    emitter.emit(key, { scope, payload }),
  );

  return [dispatch, emitter];
};
