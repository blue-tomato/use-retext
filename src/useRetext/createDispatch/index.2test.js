import createDispatch from '.';

const { store } = global;

describe('createDispatch', () => {
  it('dispatch works', () => {
    const [dispatch, emitter] = createDispatch(store.action);
    const test = ({ scope, payload }) => expect({ scope, payload }).toMatchSnapshot();

    emitter.on('increment', test);
    emitter.on('toggle', test);
    emitter.on('setOpen', test);

    dispatch.increment();
    dispatch.sideMenu.increment();
    dispatch.sideMenu.toggle();
    dispatch.sideMenu.child.toggle();
  });
});
