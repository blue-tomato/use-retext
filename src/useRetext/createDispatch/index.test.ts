import createDispatch from '.';

const { store } = global;

interface Event {
  scope: string;
  payload: any;
}

describe('createDispatch', () => {
  it('dispatch works', () => {
    const [dispatch, emitter] = createDispatch(store.action);
    const test = ({ scope, payload }: Event) => expect({ scope, payload }).toMatchSnapshot();

    emitter.on('increment', test);
    emitter.on('toggle', test);
    emitter.on('setOpen', test);

    dispatch.increment();
    dispatch.sideMenu.increment();
    dispatch.sideMenu.toggle();
    dispatch.sideMenu.child.toggle();
  });
});
