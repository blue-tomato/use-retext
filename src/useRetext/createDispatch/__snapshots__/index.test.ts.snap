// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDispatch dispatch works 1`] = `
Object {
  "payload": undefined,
  "scope": "",
}
`;

exports[`createDispatch dispatch works 2`] = `
Object {
  "payload": undefined,
  "scope": "sideMenu",
}
`;

exports[`createDispatch dispatch works 3`] = `
Object {
  "payload": undefined,
  "scope": "sideMenu",
}
`;

exports[`createDispatch dispatch works 4`] = `
Object {
  "payload": undefined,
  "scope": "sideMenu.child",
}
`;
